pip/pip-338.md

```diff
@@ -0,0 +1,172 @@
+
+Also, Introduce a broker-level property to determine the appropriate listener for lookup calls if listenerName is missing from the client request. This can help deal with the side effects of not having listerName on the client side when multiple advertisedListeners are being used.
+
+Also, this can help to determine the broker service URL that needs to be returned in case of lookup redirects and also for using Pulsar admin API calls from outside the network.
+
+This will help in having a more transparent listener identifier for internal and external communications.
+
+
+# Design & Implementation Details
+## Approach
+**1. Fix Code Flow - Inconsistency with listener usage with protocols other than pulsar and pulsar+ssl** 
+
+Currently, lookup result return the broker service url corresponding to the internal listerner in case of http or https protocol urls. To address the issue in the workflow, it is necessary to include all broker addresses associated with the listener name, rather than solely the service URL. There can be a check if all broker addresses corresponding to the given listener is null, then use the default listener(discussed below)
+
+**2. Only return the service URL corresponding to the lookupListener**
+
+This approach introduces one new configuration `lookupListenerName` in the broker.conf. The `lookupListenerName` is used to specify which service URL should be returned to the clients from the lookup requests or redirects. `lookupListenerName` should be present in the `advertisedListeners` list.
+Users can set up the `lookupListenerName`. for example:
+
+```
+existing configs:
+advertisedListeners=internal:pulsar://xyz-broker:6660,internal:pulsar+ssl://192.168.1.11:6651,external:http://192.168.1.11:8080
+internalListenerName=internal
+
+new config required:
+lookupListenerName=external
+```
+
+In the approach, when the client sends a lookup request to the broker to get the owner broker without the listnerName, the broker only returns one service URL that is with the same listener name for which the default value of the lookupListenerName is given in the broker configuration. Therefore, we must allow the client to get the corresponding service URL with the same advertised listener name as present in the broker config.
+
+This approach's purpose is to keep client-side simple and handle situations when not having listenerName information can lead to breaking scenarios. By having the default value, clients should be able to connect with the broker even if the request lands to the broker who is not the owner and redirects to some other broker.
+
+## Changes
+1. Fix the lookup result formation with all broker addresses rather than only service url
+   in  NamespaceService.java in following methods:
+```java
+- CompletableFuture<Optional<LookupResult>> findBrokerServiceUrl(NamespaceBundle bundle, LookupOptions options)
```

Comment:
a LookupResult is not needed for addressing the Pulsar Admin API's redirects. Please check the previous comment about PulsarWebResource.maybeRedirectToBroker.

Reply:
The `/lookup/v2/topic` API and many other places do not use the above approach (`maybeRedirectToBroker`) but rather, rely on the lookup result response..

https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/lookup/TopicLookupBase.java#L107

https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java#L930

to name some.

The `maybeRedirectToBroker` method is only used in the API `cluster/brokerid/ownedNamespaces` and even inside this `maybeRedirectToBroker`, today, uses the internal listener's http url for redirect. The goal is to have a common method which figures out the correct redirect url in PulsarWebResource.java itself. Maybe it can be this same exact method. We will send a draft PR with changes for early review if needed by tomorrow.

Reply:
Yes, there's a need to make things consistent in the implementation as you have suggested.